File: 카카오/2020카카오블라인드1차_2.cpp
#include<bits/stdc++.h>

using namespace std;

bool check(string p) {
    int count = 0;

    for (int i = 0; i < p.size(); i++) {
        p[i] == '(' ? count++ : count--;
        if (count < 0)
            return false;
    }
    return true;
}

string solution(string p) {
    if (p == "")
        return "";
    if (check(p) == true)
        return p;

    int i, count = 0;
    for (i = 0; i < p.size(); i++) {
        p[i] == '(' ? count++ : count--;
        if (count == 0)
            break;
    }

    string u = p.substr(0, i + 1);
    string v = p.substr(i + 1);

    if (check(u) == true) {
        return u + solution(v);
    }
    else {
        string tmp = "(" + solution(v) + ")";
        u = u.substr(1, u.size() - 2);
        
        for (int i = 0; i < u.size(); i++)
            u[i] = u[i] == '(' ? ')' : '(';

        tmp += u;
        return tmp;
    }
}



int main() {
    cout << solution("()))((()");
}

File: 카카오/2020카카오블라인드1차_2.java
class Solution {
    public String solution(String p) {
        if (p.isEmpty() || isCorrect(p))
            return p;

        String u = "", v = "";

        int count = 0;
        for (int i = 0; i < p.length(); i++) {
            count += p.charAt(i) == '(' ? 1 : -1;
            if (count == 0) {
                u = p.substring(0, i + 1);
                v = p.substring(i + 1);
                break;
            }
        }

        if (isCorrect(u))
            return u + solution(v);

        String tmp = "(" + solution(v) + ")";

        for (int i = 1; i < u.length() - 1; i++) {
             tmp += u.charAt(i) == ')'? '(' : ')';
        }
        return tmp;
    }

    private boolean isCorrect(String str) {
        int count = 0;
        for (int i = 0; i < str.length(); i++) {
            count += str.charAt(i) == '(' ? 1 : -1;
            if (count < 0)
                return false;
        }
        return true;
    }
}